apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore.Tests/Trace/CloudTraceExtensionTest.cs

```diff
@@ -19,33 +19,74 @@
 using System;
 using Xunit;
 
+using static Google.Cloud.Diagnostics.AspNetCore.CloudTraceExtension;
+
 namespace Google.Cloud.Diagnostics.AspNetCore.Tests
 {
     public class CloudTraceExtensionTest
     {
-        private const string TraceId = "105445aa7843bc8bf206b12000100f00";
-        private const ulong SpanId = 81237123;
+        private const string _traceId = "105445aa7843bc8bf206b12000100f00";
+        private const ulong _spanId = 81237123;
 
-        [Fact]
-        public void CreateTraceHeaderContext()
+        /// <summary>
+        /// Create an <see cref="IServiceProvider"/> to be used during tests of 
+        /// <see cref="CloudTraceExtension.CreateTraceHeaderContext(IServiceProvider)"/>.
+        /// It will set up an <see cref="IHttpContextAccessor"/> with the passed in trace header,
+        /// a <see cref="ShouldTraceRequest"/> with the passed in should trace function and a
+        /// <see cref="TraceIdFactory"/>.
+        /// </summary>
+        private IServiceProvider CreateProviderForTraceHeaderContext(
+            string traceHeader, Func<HttpRequest, bool?> shouldTraceFunc = null)
         {
-            string header = $"{TraceId}/{SpanId};o=1";
-
             var context = new DefaultHttpContext();
             var request = new DefaultHttpRequest(context);
-            request.Headers[TraceHeaderContext.TraceHeader] = header;
+            request.Headers[TraceHeaderContext.TraceHeader] = traceHeader;
 
             var accessor = new HttpContextAccessor();
             accessor.HttpContext = context;
-            
+
+            var traceIdFactory = TraceIdFactory.Create();
+
             Mock<IServiceProvider> mockProvider = new Mock<IServiceProvider>();
             mockProvider.Setup(p => p.GetService(typeof(IHttpContextAccessor))).Returns(accessor);
+            mockProvider.Setup(p => p.GetService(typeof(ShouldTraceRequest))).Returns(
+                new ShouldTraceRequest { ShouldTrace = shouldTraceFunc });
+            mockProvider.Setup(p => p.GetService(typeof(TraceIdFactory))).Returns(traceIdFactory);
+            return mockProvider.Object;
+        }
 
-            var headerContext = CloudTraceExtension.CreateTraceHeaderContext(mockProvider.Object);
+        [Fact]
+        public void CreateTraceHeaderContext()
+        {
+            var header = $"{_traceId}/{_spanId};o=1";
+            var provider = CreateProviderForTraceHeaderContext(header);
+            var headerContext = CloudTraceExtension.CreateTraceHeaderContext(provider);
             Assert.Equal(TraceHeaderContext.FromHeader(header).ToString(), headerContext.ToString());
         }
 
         [Fact]
+        public void CreateTraceHeaderContext_ForceTrace_False() =>
```

Comment:
You can just use a parameterized test for this. See [here](https://github.com/GoogleCloudPlatform/google-cloud-dotnet/blob/master/apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.Tests/BucketValidationTest.cs#L22) for an example.